### PR TITLE
Run delta-pass with child process

### PR DESCRIPTION
**Pull request by rushiraj7677**
Add a command line option that causes us to do the transformation part of the delta loop in a sub-process, so if a transformation crashes (this happens and it’s super annoying) it doesn’t take out the entire llvm-reduce

### CODE REVIEW

File: llvm/test/tools/llvm-reduce/remove-alias.ll [hunk 1]
```diff
@@ -1,6 +1,8 @@
 ; RUN: llvm-reduce --test FileCheck --test-arg --check-prefixes=CHECK-INTERESTINGNESS --test-arg %s --test-arg --input-file %s -o %t
 ; RUN: FileCheck --check-prefixes=CHECK-FINAL --input-file=%t %s
 
+; RUN: llvm-reduce --run-delta-in-child --test FileCheck --test-arg --check-prefixes=CHECK-INTERESTINGNESS --test-arg %s --test-arg --input-file %s -o %t
```
Comment (vivekvpandya) [inline llvm/test/tools/llvm-reduce/remove-alias.ll:4]: @rushiraj7677 our new option will not work in windows. So these tests will fail on windows machine. 
Perhaps we need to think about how we can support this in non posix environment.

**Comment by vivekvpandya** [comment]
Just to note for other reviewers we have tested this by setting default value for `--run-delta-in-child` to true. Also we added explicit abort to a delta pass for testing and when such thing happens in child process , parent reports name of that delta pass.

**Comment by vivekvpandya** [comment]
@regehr @arsenm kindly review!

**Comment by arsenm** [comment]
I think this is a bad direction to go in. If there are crashes in the reductions themselves, we should just fix them. This is just papering over real issues

File: llvm/test/tools/llvm-reduce/run-all-delta-in-child.ll [hunk 2]
```diff
@@ -0,0 +1,53 @@
+; UNSUPPORTED: target={{.-windows.}}
+; RUN: llvm-reduce --run-delta-in-child --test FileCheck --test-arg --check-prefixes=CHECK-INTERESTINGNESS --test-arg %s --test-arg --input-file %s -o %t
+; RUN: FileCheck --check-prefixes=CHECK-FINAL --input-file=%t %s
+
```
Comment (arsenm) [inline llvm/test/tools/llvm-reduce/run-all-delta-in-child.ll:4]: So does this test demonstrate a crash in a reduction? I'm confused about what this is showing

**Comment by regehr** [comment]
> I think this is a bad direction to go in. If there are crashes in the reductions themselves, we should just fix them. This is just papering over real issues

but papering over real issues is useful. for example if I'm reducing something that triggers a crash in simplifyCFG, then llvm-reduce is going to tend to crash when it runs simplifyCFG. if we have a super easy workaround for this, we'll get a reduced test case faster, and then we'll get the bug fixed faster. 

as long as this is turned off by default, we're only papering over problems when we need to do so. I'm in favor of a patch that does this.

**Comment by arsenm** [comment]
> > I think this is a bad direction to go in. If there are crashes in the reductions themselves, we should just fix them. This is just papering over real issues
> 
> but papering over real issues is useful. for example if I'm reducing something that triggers a crash in simplifyCFG, then llvm-reduce is going to tend to crash when it runs simplifyCFG. if we have a super easy workaround for this, we'll get a reduced test case faster, and then we'll get the bug fixed faster.

But how often is this actually happening? Fixing crashes on some unhandled IR is easy, and I can't recall ever encountering a reduction crashing 

> 
> as long as this is turned off by default, we're only papering over problems when we need to do so. I'm in favor of a patch that does this.

**Comment by regehr** [comment]
> But how often is this actually happening? Fixing crashes on some unhandled IR is easy, and I can't recall ever encountering a reduction crashing

I only have my own experience to draw on. llvm-reduce crashing out before it can reduce something for me has happened a number of times, and this has been frustrating.

if you want to block this patch then I guess block the patch, but it seems simple and low cost and it solves an actual problem that has happened to me, and I assume to others. I don't see the problem here.

**Pull request closed without merging**